packages/api/internal/cache/instance/instance.go

```diff
@@ -147,125 +148,44 @@ func (i *InstanceInfo) SetExpired() {
 	i.SetEndTime(time.Now())
 }
 
-type InstanceCache struct {
-	reservations *ReservationCache
-	pausing      *smap.Map[*InstanceInfo]
-
-	cache          *lifecycleCache[*InstanceInfo]
-	insertInstance func(data *InstanceInfo, created bool) error
-
-	sandboxCounter metric.Int64UpDownCounter
-	createdCounter metric.Int64Counter
+func (i *InstanceInfo) PauseStarted() bool {
+	i.mu.RLock()
+	defer i.mu.RUnlock()
 
-	mu sync.Mutex
+	return i.pausingState != pausingNotStarted
 }
 
-func NewCache(
-	ctx context.Context,
-	meterProvider metric.MeterProvider,
-	insertInstance func(data *InstanceInfo, created bool) error,
-	deleteInstance func(data *InstanceInfo) error,
-) *InstanceCache {
-	// We will need to either use Redis or Consul's KV for storing active sandboxes to keep everything in sync,
-	// right now we load them from Orchestrator
-	cache := newLifecycleCache[*InstanceInfo]()
-
-	meter := meterProvider.Meter("api.cache.sandbox")
-	sandboxCounter, err := telemetry.GetUpDownCounter(meter, telemetry.SandboxCountMeterName)
-	if err != nil {
-		zap.L().Error("error getting counter", zap.Error(err))
-	}
-
-	createdCounter, err := telemetry.GetCounter(meter, telemetry.SandboxCreateMeterName)
-	if err != nil {
-		zap.L().Error("error getting counter", zap.Error(err))
-	}
-
-	instanceCache := &InstanceCache{
-		cache:          cache,
-		insertInstance: insertInstance,
-		sandboxCounter: sandboxCounter,
-		createdCounter: createdCounter,
-		reservations:   NewReservationCache(),
-		pausing:        smap.New[*InstanceInfo](),
-	}
-
-	cache.OnEviction(func(ctx context.Context, instanceInfo *InstanceInfo) {
-		err := deleteInstance(instanceInfo)
-		if err != nil {
-			zap.L().Error("Error deleting instance", zap.Error(err))
-		}
-
-		instanceCache.UpdateCounters(ctx, instanceInfo, -1, false)
-	})
-
-	go cache.Start(ctx)
-
-	return instanceCache
-}
+func (i *InstanceInfo) StartPausing() {
+	i.mu.Lock()
+	defer i.mu.Unlock()
 
-func (c *InstanceCache) Len() int {
-	return c.cache.Len()
+	i.pausingState = pausingInProgress
 }
 
-func (c *InstanceCache) Set(key string, value *InstanceInfo, created bool) {
-	inserted := c.cache.SetIfAbsent(key, value)
-	if inserted {
-		go func() {
-			err := c.insertInstance(value, created)
-			if err != nil {
-				zap.L().Error("error inserting instance", zap.Error(err))
-			}
-		}()
+func (i *InstanceInfo) WaitForPausing(ctx context.Context) error {
+	if !i.PauseStarted() {
+		return fmt.Errorf("sandbox isn't pausing")
 	}
-}
 
-func (c *InstanceCache) MarkAsPausing(instanceInfo *InstanceInfo) {
```

Comment:
refactored this, now it all lives on the `InstanceInfo` and was basically moved into one boolean